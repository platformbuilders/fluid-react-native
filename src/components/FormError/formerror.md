
# FormError ❗
![WhatsApp Image 2020-05-05 at 12 57 39](https://user-images.githubusercontent.com/44801113/81088031-d076a680-8ed0-11ea-8200-2b6e0df7220f.jpeg)


```js
import React from 'react';
import { FormError } from '@platformbuilders/react-native-ui';

const Home: React.FC = () => {
  return (
      <FormError
        error="here is error message"
        centered
      />
  );
}

export default Home;
```


| Name  | Required | Type | Default Value | Description 
| ------------- | ------------- | ------------- |------------- |------------- |
| centered | false | boolean | false | center string |
| error | false | string | ' ' |  |
| children | false | React.Children |  | children  |
| style | false | CSSProperties |  |   |


