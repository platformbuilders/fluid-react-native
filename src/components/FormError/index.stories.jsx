import { storiesOf } from '@storybook/react-native';
import formerrorMarkdown from './formerror.md';

storiesOf('FormError', module)
  .addParameters({
    readme: {
      content: formerrorMarkdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => {});
