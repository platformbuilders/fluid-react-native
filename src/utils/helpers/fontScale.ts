import { RFValue } from 'react-native-responsive-fontsize';

export const scaledFontSize = (
  size: number | string,
  baselineFactor?: 280,
): number => {
  const value = typeof size === 'string' ? parseInt(size, 10) : size;
  return RFValue(value, baselineFactor);
};
