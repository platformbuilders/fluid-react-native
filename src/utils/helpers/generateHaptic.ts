import Haptic, { HapticFeedbackType } from 'react-native-haptic';
import { isIOS } from './isIOS';

export const generateHaptic = (haptic: HapticFeedbackType = 'impact'): void => {
  if (isIOS()) {
    Haptic.generate(haptic);
  }
};
