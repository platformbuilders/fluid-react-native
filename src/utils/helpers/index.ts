export {
  currencyParser,
  parseToThousands,
  removeWhiteSpaces,
  toOnlyNumbers,
  getTheme,
  ifStyle,
  switchStyle,
} from '@platformbuilders/helpers';
export * from './generateHaptic';
export * from './getByTypographyVariant';
export * from './fontScale';
export * from './shadow';
