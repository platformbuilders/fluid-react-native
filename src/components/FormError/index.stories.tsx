import type { Meta } from '@storybook/react';
import FormError from '.';

const meta: Meta<typeof FormError> = {
  title: 'FormError',
  component: FormError,
  args: {
    accessibility: 'Texto de erro',
    centered: false,
    id: 'form-error-id',
    error: 'Mensagem de erro',
    large: false,
    style: {},
  },
  parameters: {
    notes:
      'Esse componente tem como função auxiliar na padronização de exibição de erros.',
  },
};

export default meta;

export const Default = {};
