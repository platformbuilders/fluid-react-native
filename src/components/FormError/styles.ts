import styled from 'styled-components/native';
import { moderateScale } from 'react-native-size-matters';
import Typography from '../Typography';
import { getTheme, ifStyle } from '../../utils/helpers';

const isCentered = ifStyle('centered');
const isLarge = ifStyle('large');
const failure = getTheme('failure');

type ErrorTextProps = {
  centered: boolean;
  large: boolean;
  accessibilityLabel: string;
  testID: string;
};

export const ErrorText = styled(Typography).attrs((props: ErrorTextProps) => ({
  variant: isLarge('footnote', 'footnote')(props),
}))<ErrorTextProps>`
  color: ${failure};
  text-align: ${isCentered('center', 'left')};
  margin-top: ${isLarge(0, moderateScale(2))}px;
`;
