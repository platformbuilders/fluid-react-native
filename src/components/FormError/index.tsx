import React, { FC } from 'react';
import { ErrorText } from './styles';

const warnBoolean = (): void =>
  // eslint-disable-next-line no-console
  console.warn(
    "@platformbuilders/react-native-ui: received a truthy boolean error instead of string, which won't be rendered.",
  );

type Props = {
  centered?: boolean;
  large?: boolean;
  error?: string | string[] | boolean;
  id?: string;
  accessibility?: string;
};

const FormError: FC<Props> = ({
  id = 'form_error',
  accessibility,
  error = '',
  centered = false,
  large = false,
  children,
  ...rest
}) => {
  if (error && typeof error === 'boolean') warnBoolean();
  return (
    <>
      {children}
      {error && typeof error === 'string' ? (
        <ErrorText
          centered={centered}
          large={large}
          testID={`error_${id}`}
          accessibility={`Erro ${accessibility || error}`}
          accessibilityLabel={`Erro ${accessibility || error}`}
          {...rest}
        >
          {error}
        </ErrorText>
      ) : null}
    </>
  );
};

export default FormError;
