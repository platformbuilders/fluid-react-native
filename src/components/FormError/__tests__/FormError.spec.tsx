import 'jest';
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import FormError from '..';
import { theme } from '../../../test/helpers';

describe('<FormError />', () => {
  it('should render the form error component', () => {
    const wrapper = renderer.create(
      <ThemeProvider theme={theme}>
        <FormError id="" accessibility="" error="testing error" large />
      </ThemeProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
