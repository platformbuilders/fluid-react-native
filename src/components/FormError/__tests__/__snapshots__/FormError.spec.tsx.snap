// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormError /> should render the form error component 1`] = `
<Text
  accessibilityLabel="Erro testing error"
  centered={false}
  large={true}
  style={
    Array [
      Object {
        "color": "#212121",
        "fontSize": 26,
        "lineHeight": 35,
      },
      Object {
        "color": "#cc0000",
        "marginTop": 0,
        "textAlign": "left",
      },
    ]
  }
  testID="error_"
  variant="footnote"
>
  testing error
</Text>
`;
